File: video-service/src/client/java/com/boclips/videos/service/client/internal/resources/PlaybackResource.java
package com.boclips.videos.service.client.internal.resources;

import com.boclips.videos.service.client.Playback;
import com.boclips.videos.service.client.PlaybackLinks;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.Duration;

@Builder
@AllArgsConstructor
@NoArgsConstructor
@Data
public class PlaybackResource {
    private PlaybackLinksResource _links;
    private String id;
    private String thumbnailUrl;
    private Duration duration;
    private String referenceId;

    public Playback toPlayback() {
        return Playback.builder()
                .playbackId(id)
                .duration(duration)
                .thumbnailUrl(thumbnailUrl)
                .referenceId(referenceId)
                .links(new PlaybackLinks(
                        _links.getCreatePlaybackEvent()
                ))
                .build();
    }
}

File: video-service/src/main/kotlin/com/boclips/videos/service/presentation/video/playback/PlaybackResource.kt
package com.boclips.videos.service.presentation.video.playback

import java.time.Duration

sealed class PlaybackResource {
    abstract var id: String?
    abstract var thumbnailUrl: String?
    abstract var duration: Duration?
    abstract val type: String
}

data class StreamPlaybackResource(
    override val type: String = "STREAM",
    override var id: String?,
    override var thumbnailUrl: String?,
    override var duration: Duration?,
    val streamUrl: String,
    val referenceId: String
) : PlaybackResource()

data class YoutubePlaybackResource(
    override val type: String = "YOUTUBE",
    override var id: String?,
    override var thumbnailUrl: String?,
    override var duration: Duration?
) : PlaybackResource()
